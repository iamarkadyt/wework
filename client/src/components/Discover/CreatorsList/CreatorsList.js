import React, { Fragment } from 'react'
import ListNode from '../ListNode/ListNode'
import { withEither, withAdded } from '../../../hocs/conditionalRendering'
import { compose } from 'recompose'
import FBSpinner from '../../FBSpinner/FBSpinner'
import { arrayOf, func } from 'prop-types'
import { discoverListNodeType } from '../../../types/index'


const CreatorsList = ({
    followAPerson,
    list
}) => (
        <Fragment>
            <h3>Add these people to your feed:</h3>
            <div>
                {list.map(item => {
                    return (
                        <ListNode
                            key={item._id}
                            {...item}
                            followAPerson={followAPerson} />
                    )
                })}
            </div>
        </Fragment>
    )

CreatorsList.propTypes = {
  list: arrayOf(discoverListNodeType).isRequired,
  followAPerson: func.isRequired
}


const isLoadingFn = ({ list }) => !list
const isEmptyFn = ({ list }) => list.length === 0
const EmptyList = () => <p className="Discover-message">Nothing for you right now! Come back later!</p>
const withCondRenderings = compose(
  withEither(isLoadingFn, FBSpinner),
  withAdded(isEmptyFn, EmptyList), 
)
const CreatorsListWithCondRenderings = withCondRenderings(CreatorsList)


export const _UnconnectedCreatorsList = CreatorsListWithCondRenderings
export default CreatorsListWithCondRenderings
