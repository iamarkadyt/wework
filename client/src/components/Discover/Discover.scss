@import 'variables.scss';
.Discover-container {
    position: sticky;
    box-shadow: $box_shadow_main;
    background-color: $color-bg-secondary;
    color: $color-font-main;
    border-radius: .1rem;
    padding: 1rem;
    width: 100%;
    top: 4rem;
    .ListNode-container {
        margin-top: 1rem;
        display: grid;
        grid-template-columns: 3rem auto 1.4rem;
        grid-column-gap: .5rem;
        align-items: center;
        .ListNode-img {
            width: 3rem;
            border-radius: 3rem;
            height: 3rem;
        }
        .ListNode-name {
            p {
                &:first-child {
                    font-size: 1.2rem;
                    font-weight: bold;
                }
            }
        }
    }
    .Discover-message {
        background-color: $color-bg-main;
        color: $color-font-secondary;
        padding: .3rem 1rem;
        margin-top: .7rem;
        border-radius: .1rem;
    }
}