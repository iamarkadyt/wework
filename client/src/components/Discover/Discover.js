import React, { Component } from 'react'
import { connect } from 'react-redux'
import { arrayOf, func } from 'prop-types'

import './Discover.scss'
import { followAPerson, fetchUsersStats, fetchDiscoverContent } from '../../state/actions/userActions'
import { fetchPosts } from '../../state/actions/postsActions'
import CreatorsList from './CreatorsList/CreatorsList'
import { discoverListNodeType } from '../../types/index'


class Discover extends Component {
    followAPerson = (_id, callback) => {
        const {
            followAPerson,
            fetchUsersStats,
            fetchPosts,
            fetchDiscoverContent
        } = this.props

        followAPerson(_id, () => {
            fetchDiscoverContent(5)
            fetchUsersStats()
            fetchPosts(false)
            if (callback) callback()
        })
    }

    render() {
        return (
            <div className="Discover-container">
                <CreatorsList
                    list={this.props.list}
                    followAPerson={this.followAPerson} />
            </div>
        )
    }

    componentDidMount() {
        this.props.fetchDiscoverContent(5)
    }
}

Discover.propTypes = {
  list: arrayOf(discoverListNodeType),
  followAPerson: func.isRequired,
  fetchUsersStats: func.isRequired,
  fetchDiscoverContent: func.isRequired,
  fetchPosts: func.isRequired
}


export const _UnconnectedDiscover = Discover
export default connect(state => ({ list: state.user.discoverList }), {
    followAPerson,
    fetchUsersStats,
    fetchPosts,
    fetchDiscoverContent
})(Discover)
