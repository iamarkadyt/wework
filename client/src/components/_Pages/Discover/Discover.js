import React from 'react'
import './Discover.css'
import mockProfileObject from '../../../mocks/profile'
import mockPosts from '../../../mocks/posts'
import ava_placeholder from '../../../images/avatar_placeholder.png'
import { Link } from 'react-router-dom'

const discover = () => {
    let profiles = new Array(10).fill(mockProfileObject, 0, 10)

    return <div className="Discover-container">
        <input type="text" placeholder="Search people or posts, by name, text, skills, etc..." className="search" />
        <div className="content">
            <h2>People</h2>
            <h2>Posts</h2>
            <div className="people">
                {profiles.map(({ image, user, title, status, location, company, handle }, idx) => {
                    return <div className="node" key={`discover-node-profile-${idx}`}>
                        <img src={image || ava_placeholder} alt='' />
                        <div className="text">
                            <p>{user.name}</p>
                            <p>{title} at {company}</p>
                            <p>{status}, {location}</p>
                        </div>
                        <Link to='/profile'><span>View Profile</span></Link>
                    </div>
                })}
            </div>
            <div className="posts">
                {mockPosts.map(({ user, text, comments, likes }, idx) => {
                    return <div className="node" key={`discover-node-post-${idx}`}>
                        <p>{user.name} said:</p>
                        <p>{text}</p>
                        <p>
                            <a href="#">{likes.length} [LIKE!]</a>
                            <Link to='/comments'>{comments.length} comments</Link>
                        </p>
                    </div>
                })}
            </div>
        </div>
    </div>
}

export default discover