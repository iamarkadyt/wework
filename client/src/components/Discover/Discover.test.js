/**
 * Contract:
 *
 * What my component renders? What do I pass to components I render?
 *   - CreatorsListWithCondRenderings taking list and func as props
 *
 * What do I do with the props I receive, what effect do they have?
 * - 4 functions are combined in 1 convenience function and passed on to the CreatorsList
 * - List is passed on to CreatorsList
 *
 * What do I do with the state? How do I update it?
 * - N/A
 *
 * Lifecycle hooks effects? Side effects?
 * - componentDidMount calls discover content fetch 
 *
 * Interaction with the component?
 * - N/A
 *
 * Context?
 * - N/A
 *
 * Public ref API?
 * - N/A
 *
 *
 *
 */


import React from 'react'
import {shallow} from 'enzyme'
import { _UnconnectedDiscover as Discover } from './Discover'
import cloneDeep from 'lodash.clonedeep'
import CreatorsList from './CreatorsList/CreatorsList'

describe('Discover', () => {
  let props
  let mountedDiscover

  const getMockProps = () => {
    const actions = {
      followAPerson: jest.fn((it, cb) => cb()),
      fetchUsersStats: jest.fn(),
      fetchPosts: jest.fn(),
      fetchDiscoverContent: jest.fn()
    }

    const listElem = {
      _id: 'ab109a',
      name: 'John Doe',
      avatar: '//gravatar.com/12345',
      title: 'Web Developer',
      company: 'Some Company'
    }

    return cloneDeep({
      ...actions,
      list: [listElem]
    })
  }

  const discoverComp = () => {
    if(!mountedDiscover) {
      mountedDiscover = shallow(<Discover {...props} />)
    }
    return mountedDiscover
  }

  beforeEach(() => {
    props = getMockProps()
    mountedDiscover = undefined
  })

  it('local followAPerson operates as expected', () => {
    const cb = jest.fn()
    const id = 'id123'
    // this is internal followAPerson func!
    discoverComp().instance().followAPerson(id, cb)

    expect(cb).toHaveBeenCalled()
    expect(props.followAPerson).toHaveBeenCalledWith(id, expect.any(Function))
    expect(props.fetchDiscoverContent).toHaveBeenCalledWith(5)
    expect(props.fetchUsersStats).toHaveBeenCalled()
    expect(props.fetchPosts).toHaveBeenCalledWith(false)
  })

  describe('side effects', () => {
    it('calls fetchDiscoverContent() on componentDidMount', () => {
      discoverComp()
      expect(props.fetchDiscoverContent).toHaveBeenCalled()
    })
  })

  describe('props passing', () => {
    it('CreatorsList receives correct props', () => {
      const expectedProps = {
        list: props.list, 
        followAPerson: discoverComp().instance().followAPerson
      }
      expect(discoverComp().find(CreatorsList).props()).toStrictEqual(expectedProps)
    })
  })

  it('passes a snapshot test', () => {
    expect(discoverComp()).toMatchSnapshot()
  })
})
