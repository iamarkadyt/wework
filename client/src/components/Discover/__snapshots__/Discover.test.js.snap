// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discover passes a snapshot test 1`] = `
<div
  className="Discover-container"
>
  <Component
    followAPerson={[Function]}
    list={
      Array [
        Object {
          "_id": "ab109a",
          "avatar": "//gravatar.com/12345",
          "company": "Some Company",
          "name": "John Doe",
          "title": "Web Developer",
        },
      ]
    }
  />
</div>
`;
